### PR TITLE
Re-add --rust-target option to replace --unstable-rust

### PR DESCRIPTION
**Pull request by tmfink**
Re-apply commit. Addresses #832 

Instead of specifying whether or not to use stable, specify the Rust
release to support (one of several stable/beta releases or nightly).
The `--unstable-rust` option is still accepted and implies nightly.

The definitions of `RustTarget` and `RustFeatures` are created with
macros.

For each test that uses unions, there is a version that uses the latest stable
and 1.0.

### CODE REVIEW

**Comment by tmfink** [comment]
Since the last pull request, I removed unnecessary lint suppressions and warn when using the `--unstable--rust` option.

**Comment by photoszzt** [comment]
The failure is related to https://github.com/rust-lang/rust/issues/32836.

**Comment by tmfink** [comment]
It looks like this bug is not from my code, but exists in master. I modified `tests/headers/issue-493.hpp` to use "untagged_unions" rust by adding `// bindgen-flags: --unstable-rust`. When running `cargo test` from `tests/expectations` with rust 1.19, I get:

~~~
error: unions with non-`Copy` fields are unstable (see issue #32836)
  --> tests/issue-493.rs:47:1
   |
47 | / pub union basic_string___ulx {
48 | |     pub __lx: basic_string___long,
49 | |     pub __lxx: basic_string___short,
50 | | }
   | |_^

error: unions with non-`Copy` fields are unstable (see issue #32836)
  --> tests/issue-493.rs:72:1
   |
72 | / pub union basic_string___rep__bindgen_ty_1 {
73 | |     pub __l: basic_string___long,
74 | |     pub __s: basic_string___short,
75 | |     pub __r: basic_string___raw,
76 | | }
   | |_^
~~~

**Comment by tmfink** [comment]
I commented out parts of the headers that are causing the failures. These tests failed when using the Rust `union` instead of bindgen unions.

**Comment by photoszzt** [comment]
I think the problem is for the field cannot derive Copy the behavior is not standardized. You would need to add the feature gate in the header. @fitzgen maybe fall back to generate the BingenUnion when the field cannot be Copy?

**Comment by tmfink** [comment]
I filed issue #895 for the pre-existing error

File: tests/headers/16-byte-alignment.h [hunk 1]
```diff
@@ -27,7 +26,10 @@ struct rte_ipv6_tuple {
         };
 };
 
+// TODO(tmfink) uncomment once test passes
+#if 0
```
Comment (fitzgen) [inline tests/headers/16-byte-alignment.h]: Thanks for leaving them uncommented in the non-`union` cases so we don't lose test coverage. That is essential to landing this.

**Comment by fitzgen** [comment]
I'm hesitant to land this without fixing #895 at the same time (or at least before the next release), as it will expose that bug to many more people than were exposed to it before.

**Comment by fitzgen** [comment]
@tmfink do you think you could take a crack at #895? I expect it shouldn't be too difficult. Essentially, we need a new method on `ir::Comp` that checks if all of its fields `can_derive_copy` (and bases? I think `union`s can only inherit from int types in C++ but I'd need to go double check in the standard) and then check this method in `codegen` where we check the rust target features for unions.

Something like this:

```rust
fn can_be_rust_union(&self, ctx: &BindgenContext) {
    self.fields().all(|f| f.ty().can_derive_copy(ctx))
}
```

**Comment by photoszzt** [comment]
@fitzgen  What about adding a checking of the result of item.can_derive_copy(ctx) at https://github.com/rust-lang-nursery/rust-bindgen/blob/master/src/codegen/mod.rs#L1448 It should already do this check in the analysis.

**Comment by photoszzt** [comment]
The check I mean checking all the fields can derive copy

**Comment by fitzgen** [comment]
> (and bases? I think unions can only inherit from int types in C++ but I'd need to go double check in the standard)

I was actually thinking of `enum`s here. I'm even less sure about what goes on with inheritance in `union`s.

**Comment by fitzgen** [comment]
> @fitzgen What about adding a checking of the result of item.can_derive_copy(ctx) at https://github.com/rust-lang-nursery/rust-bindgen/blob/master/src/codegen/mod.rs#L1448 It should already do this check in the analysis.

Yes, this is where I was imagining the `can_be_rust_union` method would be queried.

**Comment by fitzgen** [comment]
@tmfink great, thanks! Let me know when you have something for me to look at :)

File: tests/headers/16-byte-alignment.h (hunk 1 above)
Comment (tmfink) [inline tests/headers/16-byte-alignment.h]: I left these lines disabled because of issue #908, which exist in master.

**Comment by emilio** [comment]
Can you squash the commits, please? This looks fine to me.

Thanks!

File: src/ir/comp.rs [hunk 2]
```diff
@@ -1316,6 +1317,22 @@ impl CompInfo {
...
+
+    /// Returns whether the current union can be represented as a Rust `union`
+    ///
+    /// Requirements:
+    ///     1. Current RustTarget allows for `untagged_union`
+    ///     2. Each field can derive `Copy`
+    pub fn can_be_rust_union(&self, ctx: &BindgenContext) -> bool {
+        ctx.options().rust_features().untagged_union() &&
+            self.fields().iter().all(|f|
+                match f {
```
Comment (emilio) [inline src/ir/comp.rs]: nit: Maybe `match *f` for consistency with the rest of the codebase.

**Comment by tmfink** [comment]
@emilio I squashed the commits, but it looks like the CI job failed because of an error with the setup or possibly network:
https://travis-ci.org/rust-lang-nursery/rust-bindgen/jobs/264155826

We may want to consider not spawning so many "sub-jobs". If any sub-job fails, then the main job fails.

**Comment by emilio** [comment]
@bors-servo r+

Let's see, if we see many jobs intermittently failing, we'd need to reduce the number of them I agree (they were just bumped in #901).

**Comment by fitzgen** [comment]
Note that we can restart individual jobs that fail, but I agree that isn't ideal. Still, breaking up our CI into multiple jobs shaved off more than half of our wall clock time, so I'd like to do our best to preserve it.

The intermittent there didn't look like something fundamental to having multiple jobs, just that when we have multiple jobs, we have more chances to hit an extant intermittent. If my intuition is incorrect, and this failure *is* caused by having multiple jobs, then we should fix that (of course) and investigate if regrouping jobs helps any. "Just" a matter of time, effort, and priorities :)

**Pull request merged**